panic: runtime error: index out of range in gaad.(*ADTS).excluded_channels

Hello.

I found a index out of range bug in gaad.

Please confirm.

Thanks.

reproduce code:
```
package gaad

import (
	"testing"
)

func TestFuzzCrashers(t *testing.T) {

	var crashers = []string{
		"\xff\xf10\xdc0\xb8\xad",
	}

	for _, f := range crashers {
		ParseADTS([]byte(f))
	}
}
```

Crash Log:
```
--- FAIL: TestFuzzCrashers (0.00s)
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range

goroutine 17 [running]:
testing.tRunner.func1(0xc42004d5f0)
	/usr/lib/go-1.8/src/testing/testing.go:622 +0x29d
panic(0x5337a0, 0x5fbe10)
	/usr/lib/go-1.8/src/runtime/panic.go:489 +0x2cf
github.com/Comcast/gaad.(*ADTS).excluded_channels(0xc42004d6c0, 0xc42009b301, 0x0)
	/home/karas/go/src/github.com/Comcast/gaad/aacparser.go:1832 +0x2ca
github.com/Comcast/gaad.(*ADTS).dynamic_range_info(0xc42004d6c0, 0x4, 0xb)
	/home/karas/go/src/github.com/Comcast/gaad/aacparser.go:1787 +0x3f7
github.com/Comcast/gaad.(*ADTS).extension_payload(0xc42004d6c0, 0x2, 0x4f0006, 0x0, 0x0, 0x0, 0x0)
	/home/karas/go/src/github.com/Comcast/gaad/aacparser.go:1731 +0x32b
github.com/Comcast/gaad.(*ADTS).fill_element(0xc42004d6c0, 0x6, 0x6, 0x0, 0x0)
	/home/karas/go/src/github.com/Comcast/gaad/aacparser.go:1293 +0xbf
github.com/Comcast/gaad.(*ADTS).raw_data_block(0xc42004d6c0, 0x0, 0x0)
	/home/karas/go/src/github.com/Comcast/gaad/aacparser.go:1020 +0x269
github.com/Comcast/gaad.(*ADTS).adts_frame(0xc42004d6c0, 0xc420019800, 0x7)
	/home/karas/go/src/github.com/Comcast/gaad/aacparser.go:796 +0x91
github.com/Comcast/gaad.ParseADTS(0xc4200b52a0, 0x7, 0x8, 0xc4200b52a0, 0x7, 0x8)
	/home/karas/go/src/github.com/Comcast/gaad/aacparser.go:773 +0xbe
github.com/Comcast/gaad.TestFuzzCrashers(0xc42004d5f0)
	/home/karas/go/src/github.com/Comcast/gaad/fuzzer_test.go:14 +0x90
testing.tRunner(0xc42004d5f0, 0x5647e0)
	/usr/lib/go-1.8/src/testing/testing.go:657 +0x96
created by testing.(*T).Run
	/usr/lib/go-1.8/src/testing/testing.go:697 +0x2ca
exit status 2
FAIL	github.com/Comcast/gaad	0.009s
```

--- a/aacparser.go
+++ b/aacparser.go
@@ -2152,31 +2152,7 @@ func (adts *ADTS) sbr_channel_pair_element(ext_data *sbr_extension_data, bs_amp_
 			return e, fmt.Errorf("Too many bits left, check bitstream continuity")
 		}
 
-		// TODO: Uncomment this when extensions are supported.
-		// For now this is risky when the bitstream is goofed;
-		// Sometimes is tries to allocate the world.
-		// e.Bs_extension_id = make([]uint8, 0)
-		// e.Sbr_extension = make([]*sbr_extension, 0)
-		// for i := 0; num_bits_left > 7; i++ {
-		// 	ext_id, _ := adts.reader.ReadBitsAsUInt8(2)
-		// 	e.Bs_extension_id = append(e.Bs_extension_id, ext_id)
-		// 	if e.Bs_extension_id[i] == EXTENSION_ID_PS {
-		// 		num_bits_left -= 2
-
-		// 		bits_read, ext, err := adts.sbr_extension(e.Bs_extension_id[i], num_bits_left)
-		// 		if err != nil {
-		// 			return e, err
-		// 		}
-		// 		e.Sbr_extension = append(e.Sbr_extension, ext)
-
-		// 		num_bits_left -= bits_read
-		// 		if num_bits_left < 0 {
-		// 			return e, fmt.Errorf("Error: SBR parsing overran available bits")
-		// 		}
-		// 	}
-		// }
-
-		// e.Bs_fill_bits, _ = adts.reader.ReadBitsToByteArray(num_bits_left)
+		// Extentions are currently unsupported
 		adts.reader.SkipBits(num_bits_left)
 	}
 	return e, nil

--- a/aacparser.go
+++ b/aacparser.go
@@ -22,6 +22,8 @@ import (
 	"github.com/Comcast/gaad/bitreader"
 )
 
+const MaxBitsLeft = 131072
+
 type ADTS struct {
 	Bitrate              uint32
 	ChannelConfiguration uint8
@@ -1829,7 +1831,10 @@ func (adts *ADTS) excluded_channels() (int, *excluded_channels) {
 	}
 
 	n++
-	data.Additional_excluded_chns[n-1], _ = adts.reader.ReadBitAsBool()
+
+	data.Additional_excluded_chns = make([]bool, 0)
+	additional_excluded_chn, _ := adts.reader.ReadBitAsBool()
+	data.Additional_excluded_chns = append(data.Additional_excluded_chns, additional_excluded_chn)
 	for data.Additional_excluded_chns[n-1] {
 		for i := num_excl_chan; i < num_excl_chan+7; i++ {
 			mask, _ := adts.reader.ReadBitAsBool()
@@ -1837,7 +1842,9 @@ func (adts *ADTS) excluded_channels() (int, *excluded_channels) {
 		}
 		n++
 		num_excl_chan += 7
-		data.Additional_excluded_chns[n-1], _ = adts.reader.ReadBitAsBool()
+
+		additional_excluded_chn, _ := adts.reader.ReadBitAsBool()
+		data.Additional_excluded_chns = append(data.Additional_excluded_chns, additional_excluded_chn)
 	}
 
 	return n, data
@@ -1910,6 +1917,11 @@ func (adts *ADTS) sbr_extension_data(cnt int, id_aac uint8, crc_flag bool) (int,
 	}
 
 	num_align_bits := (8*uint(cnt) - 4 - num_sbr_bits)
+
+	if 8*uint(cnt) < (4 + num_sbr_bits) {
+		return 0, data, fmt.Errorf("sbr extension payload malformed")
+	}
+
 	data.Bs_fill_bits, _ = adts.reader.ReadBitsToByteArray(num_align_bits)
 
 	return int(num_sbr_bits+num_align_bits+4) / 8, data, err
@@ -2020,6 +2032,10 @@ func (adts *ADTS) sbr_single_channel_element(ext_data *sbr_extension_data, bs_am
 		}
 
 		num_bits_left := cnt * 8
+		if num_bits_left > MaxBitsLeft {
+			return e, fmt.Errorf("Too many bits left, checl bitstream continuity")
+		}
+
 		e.Bs_extension_id = make([]uint8, 0)
 		e.Sbr_extension = make([]*sbr_extension, 0)
 		for i := 0; num_bits_left > 7; i++ {
@@ -2028,7 +2044,10 @@ func (adts *ADTS) sbr_single_channel_element(ext_data *sbr_extension_data, bs_am
 			if e.Bs_extension_id[i] == EXTENSION_ID_PS {
 				num_bits_left -= 2
 
-				bits_read, ext := adts.sbr_extension(e.Bs_extension_id[i], num_bits_left)
+				bits_read, ext, err := adts.sbr_extension(e.Bs_extension_id[i], num_bits_left)
+				if err != nil {
+					return e, err
+				}
 				e.Sbr_extension = append(e.Sbr_extension, ext)
 
 				num_bits_left -= bits_read
@@ -2128,25 +2147,37 @@ func (adts *ADTS) sbr_channel_pair_element(ext_data *sbr_extension_data, bs_amp_
 
 		// TODO: could we be a bit more graceful about this block?
 		num_bits_left := cnt * 8
-		e.Bs_extension_id = make([]uint8, 0)
-		e.Sbr_extension = make([]*sbr_extension, 0)
-		for i := 0; num_bits_left > 7; i++ {
-			ext_id, _ := adts.reader.ReadBitsAsUInt8(2)
-			e.Bs_extension_id = append(e.Bs_extension_id, ext_id)
-			if e.Bs_extension_id[i] == EXTENSION_ID_PS {
-				num_bits_left -= 2
 
-				bits_read, ext := adts.sbr_extension(e.Bs_extension_id[i], num_bits_left)
-				e.Sbr_extension = append(e.Sbr_extension, ext)
-
-				num_bits_left -= bits_read
-				if num_bits_left < 0 {
-					return e, fmt.Errorf("Error: SBR parsing overran available bits")
-				}
-			}
+		if num_bits_left > MaxBitsLeft {
+			return e, fmt.Errorf("Too many bits left, check bitstream continuity")
 		}
 
-		e.Bs_fill_bits, _ = adts.reader.ReadBitsToByteArray(num_bits_left)
+		// TODO: Uncomment this when extensions are supported.
+		// For now this is risky when the bitstream is goofed;
+		// Sometimes is tries to allocate the world.
+		// e.Bs_extension_id = make([]uint8, 0)
+		// e.Sbr_extension = make([]*sbr_extension, 0)
+		// for i := 0; num_bits_left > 7; i++ {
+		// 	ext_id, _ := adts.reader.ReadBitsAsUInt8(2)
+		// 	e.Bs_extension_id = append(e.Bs_extension_id, ext_id)
+		// 	if e.Bs_extension_id[i] == EXTENSION_ID_PS {
+		// 		num_bits_left -= 2
+
+		// 		bits_read, ext, err := adts.sbr_extension(e.Bs_extension_id[i], num_bits_left)
+		// 		if err != nil {
+		// 			return e, err
+		// 		}
+		// 		e.Sbr_extension = append(e.Sbr_extension, ext)
+
+		// 		num_bits_left -= bits_read
+		// 		if num_bits_left < 0 {
+		// 			return e, fmt.Errorf("Error: SBR parsing overran available bits")
+		// 		}
+		// 	}
+		// }
+
+		// e.Bs_fill_bits, _ = adts.reader.ReadBitsToByteArray(num_bits_left)
+		adts.reader.SkipBits(num_bits_left)
 	}
 	return e, nil
 }
@@ -2200,6 +2231,10 @@ func (adts *ADTS) sbr_channel_pair_base_element(bs_amp_res bool, ext_data *sbr_e
 
 		// TODO: could we be a bit more graceful about this block?
 		num_bits_left := cnt * 8
+		if num_bits_left > MaxBitsLeft {
+			return e, fmt.Errorf("Too many bits left, checl bitstream continuity")
+		}
+
 		e.Bs_extension_id = make([]uint8, 0)
 		e.Sbr_extension = make([]*sbr_extension, 0)
 		for i := 0; num_bits_left > 7; i++ {
@@ -2208,7 +2243,10 @@ func (adts *ADTS) sbr_channel_pair_base_element(bs_amp_res bool, ext_data *sbr_e
 			if e.Bs_extension_id[i] == EXTENSION_ID_PS {
 				num_bits_left -= 2
 
-				bits_read, ext := adts.sbr_extension(e.Bs_extension_id[i], num_bits_left)
+				bits_read, ext, err := adts.sbr_extension(e.Bs_extension_id[i], num_bits_left)
+				if err != nil {
+					return e, err
+				}
 				e.Sbr_extension = append(e.Sbr_extension, ext)
 
 				num_bits_left -= bits_read
@@ -2505,17 +2543,20 @@ func (adts *ADTS) sbr_sinusoidal_coding(ch uint8, data *sbr_sinusoidal_coding, e
 ////////////////////////////////////////////////////////////////////////////////
 // Table 8.A.1 – Syntax of sbr_extension()
 ////////////////////////////////////////////////////////////////////////////////
-func (adts *ADTS) sbr_extension(bs_extension_id uint8, num_bits_left uint) (uint, *sbr_extension) {
+func (adts *ADTS) sbr_extension(bs_extension_id uint8, num_bits_left uint) (uint, *sbr_extension, error) {
 	data := &sbr_extension{}
 	switch bs_extension_id {
 	case EXTENSION_ID_PS:
+		// he-AAC v2 - currently unsupported
 		// TODO
 		// num_bits_left -= ps_data()
+		// returning num_bits_left tells the caller that all bits have been read
+		return num_bits_left, nil, fmt.Errorf("bs_extension_id of 2 (EXTENSION_ID_PS) unsupported")
 	default:
-		data.Bs_fill_bits, _ = adts.reader.ReadBitsToByteArray(num_bits_left)
-		num_bits_left = 0
+		//data.Bs_fill_bits, _ = adts.reader.ReadBitsToByteArray(num_bits_left)
 	}
-	return num_bits_left, data
+	// returning num_bits_left tells the caller that all bits have been read
+	return num_bits_left, data, nil
 }
 
 func is_intensity(info *ics_info, group, sfb uint8) int {

--- a/aacparser.go
+++ b/aacparser.go
@@ -2033,7 +2033,7 @@ func (adts *ADTS) sbr_single_channel_element(ext_data *sbr_extension_data, bs_am
 
 		num_bits_left := cnt * 8
 		if num_bits_left > MaxBitsLeft {
-			return e, fmt.Errorf("Too many bits left, checl bitstream continuity")
+			return e, fmt.Errorf("Too many bits left, check bitstream continuity")
 		}
 
 		e.Bs_extension_id = make([]uint8, 0)
@@ -2208,7 +2208,7 @@ func (adts *ADTS) sbr_channel_pair_base_element(bs_amp_res bool, ext_data *sbr_e
 		// TODO: could we be a bit more graceful about this block?
 		num_bits_left := cnt * 8
 		if num_bits_left > MaxBitsLeft {
-			return e, fmt.Errorf("Too many bits left, checl bitstream continuity")
+			return e, fmt.Errorf("Too many bits left, check bitstream continuity")
 		}
 
 		e.Bs_extension_id = make([]uint8, 0)

--- a/aacsbrtables.go
+++ b/aacsbrtables.go
@@ -249,14 +249,16 @@ func freq_derived(data *sbr_extension_data, bs_xover_band uint8, k2 uint8) error
 	data.n[0] = data.N_low
 	data.n[1] = data.N_high
 
+	index := data.N_high + bs_xover_band + 1
+
 	// check for overflow or upper bounds
-	if data.N_high+bs_xover_band+1 < bs_xover_band || len(data.f_master) < int(data.N_high+bs_xover_band+1) {
-		return fmt.Errorf("f_tablehigh invalid index, likely malformed")
+	if index < bs_xover_band || len(data.f_master) < int(index) {
+		return fmt.Errorf("f_tablehigh invalid index: index (%d) must be between bs_xover_band (%d) and length of f_master (%d)", index, bs_xover_band, len(data.f_master))
 	}
-	data.f_tablehigh = data.f_master[bs_xover_band : data.N_high+bs_xover_band+1]
-	// index out of range [255] with len 0
+	data.f_tablehigh = data.f_master[bs_xover_band:index]
+	// check f_tablehigh bounds
 	if len(data.f_tablehigh) < int(data.N_high) {
-		return fmt.Errorf("f_tablehigh invalid index, likely malformed")
+		return fmt.Errorf("N_high index (%d) too high for length of f_tablehigh (%d)", data.N_high, len(data.f_tablehigh))
 	}
 	data.M = uint8(data.f_tablehigh[data.N_high] - data.f_tablehigh[0])
 	data.k_x = data.f_tablehigh[0]

--- a/aacsbrtables.go
+++ b/aacsbrtables.go
@@ -85,7 +85,9 @@ func derive_sbr_tables(data *sbr_extension_data, sfi uint8, bs_start_freq uint8,
 	} else {
 		freq_master(data, data.k0, data.k2, bs_freq_scale, bs_alter_scale)
 	}
-	freq_derived(data, bs_xover_band, data.k2)
+	if err := freq_derived(data, bs_xover_band, data.k2); err != nil {
+		return err
+	}
 
 	return nil
 }
@@ -239,7 +241,7 @@ func freq_master(data *sbr_extension_data, k0 uint8, k2 uint8, bs_freq_scale uin
 	data.f_master = append(data.f_master, vk1...)
 }
 
-func freq_derived(data *sbr_extension_data, bs_xover_band uint8, k2 uint8) {
+func freq_derived(data *sbr_extension_data, bs_xover_band uint8, k2 uint8) error {
 	data.N_high = data.N_master - bs_xover_band
 	data.N_low = (data.N_high >> 1) + (data.N_high - (data.N_high>>1)<<1)
 
@@ -247,7 +249,15 @@ func freq_derived(data *sbr_extension_data, bs_xover_band uint8, k2 uint8) {
 	data.n[0] = data.N_low
 	data.n[1] = data.N_high
 
+	// check for overflow or upper bounds
+	if data.N_high+bs_xover_band+1 < bs_xover_band || len(data.f_master) < int(data.N_high+bs_xover_band+1) {
+		return fmt.Errorf("f_tablehigh invalid index, likely malformed")
+	}
 	data.f_tablehigh = data.f_master[bs_xover_band : data.N_high+bs_xover_band+1]
+	// index out of range [255] with len 0
+	if len(data.f_tablehigh) < int(data.N_high) {
+		return fmt.Errorf("f_tablehigh invalid index, likely malformed")
+	}
 	data.M = uint8(data.f_tablehigh[data.N_high] - data.f_tablehigh[0])
 	data.k_x = data.f_tablehigh[0]
 
@@ -274,6 +284,8 @@ func freq_derived(data *sbr_extension_data, bs_xover_band uint8, k2 uint8) {
 		}
 		data.f_tablenoise[k] = data.f_tablelow[i]
 	}
+
+	return nil
 }
 
 // TODO: Implement freq_limiter table

